zebra-consensus/src/script.rs

```diff
@@ -12,11 +12,181 @@
+
+use crate::BoxError;
+
 /// Internal script verification service.
 ///
 /// After verification, the script future completes. State changes are handled by
 /// `BlockVerifier` or `MempoolTransactionVerifier`.
 ///
 /// `ScriptVerifier` is not yet implemented.
-#[derive(Default)]
-pub(crate) struct ScriptVerifier {}
+pub struct Verifier<ZS> {
+    state: ZS,
+    branch: ConsensusBranchId,
+}
+
+impl<ZS> Verifier<ZS> {
+    pub fn new(state: ZS, branch: ConsensusBranchId) -> Self {
+        Self { state, branch }
+    }
+}
+
+#[derive(Debug)]
+struct Request {
+    transaction: Arc<Transaction>,
+    input_index: usize,
+}
+
+impl<ZS> tower::Service<Request> for Verifier<ZS>
+where
+    ZS: tower::Service<zebra_state::Request, Response = zebra_state::Response, Error = BoxError>,
+    ZS::Future: Send + 'static,
+{
+    type Response = ();
+    type Error = BoxError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.state.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        use futures_util::FutureExt;
+
+        let input = &req.transaction.inputs()[req.input_index];
+
+        match input {
+            transparent::Input::PrevOut { outpoint, .. } => {
+                let output = self.state.call(zebra_state::Request::AwaitUtxo(*outpoint));
+                let transaction = req.transaction;
+                let branch_id = self.branch;
+                let input_index = req.input_index;
+
+                async move {
+                    let previous_output = match output.await? {
+                        zebra_state::Response::Utxo(output) => output,
+                        _ => unreachable!("AwaitUtxo always responds with Utxo"),
+                    };
+
+                    zebra_script::is_valid(
+                        transaction,
+                        branch_id,
+                        (input_index as u32, previous_output),
+                    )?;
+
+                    Ok(())
+                }
+                .boxed()
+            }
+            transparent::Input::Coinbase { .. } => unimplemented!(
+                "how should we handle verifying coinbase transactions in the script::Verifier?"
+            ),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use color_eyre::Report;
+
+    use zebra_chain::{
+        block::Block, parameters::Network, parameters::NetworkUpgrade,
+        serialization::ZcashDeserializeInto, transparent,
+    };
+    use zebra_test::transcript::{TransError, Transcript};
+
+    #[tokio::test]
+    async fn happy_path() -> Result<(), Report> {
+        happy_path_test().await
+    }
+
+    #[spandoc::spandoc]
+    async fn happy_path_test() -> Result<(), Report> {
```

Comment:
It seems like this test is actually 3-4 separate tests in the same function.
Would it help to split them up?

Reply:
its two tests, tho the second one (the initial assert) was added because I realized the rest of the test was useless and I wanted to write a short gist of why this test was doomed to fail. I don't think splitting it up is useful, I still have to read the other comments, so there might be a suggestion for how to make this actually useful that I've missed but my expectation is that this entire test needs to get tossed in the trash.